google/cloud/forseti/scanner/scanners/external_project_access_scanner.py

```diff
@@ -0,0 +1,286 @@
+                rules_file_path=self.rules,
+                snapshot_timestamp=self.snapshot_timestamp
+            )
+
+        self.rules_engine.build_rule_book(self.inventory_configs)
+
+        self._ancestries = dict()
+
+    def _output_results(self, all_violations):
+        """Output results.
+
+        Args:
+            all_violations (list): A list of violations.
+        """
+        all_violations = self._flatten_violations(all_violations)
+
+        self._output_results_to_db(all_violations)
+
+    def _find_violations(self, ancestries_by_user):
+        """Find violations in the policies.
+
+        Args:
+            ancestries_by_user (dict): The project ancestries collected
+                                               from the scanner
+        Returns:
+            list: A list of ExternalProjectAccess violations
+        """
+        all_violations = []
+        LOGGER.info('Finding project access violations...')
+
+        for user_mail, project_ancestries in ancestries_by_user.iteritems():
+            for project_ancestry in project_ancestries:
+                violations = \
+                    self.rules_engine.find_policy_violations(user_mail,
+                                                             project_ancestry)
+                all_violations.extend(violations)
+
+        return all_violations
+
+    @staticmethod
+    def _flatten_violations(violations):
+        """Flatten RuleViolations into a dict for each RuleViolation member.
+
+        Args:
+            violations (list): The RuleViolations to flatten.
+
+        Yields:
+            dict: Iterator of RuleViolations as a dict per member.
+        """
+        for violation in violations:
+
+            violation_data = {
+                'full_name': violation.full_name,
+                'member': violation.member,
+                'rule_ancestor': violation.rule_ancestor.name
+            }
+
+            yield {
+                'resource_id': violation.resource_id,
+                'resource_type': violation.resource_type,
+                'full_name': violation.full_name,
+                'rule_index': violation.rule_index,
+                'rule_name': violation.rule_name,
+                'violation_type': violation.violation_type,
+                'violation_data': violation_data,
+                'resource_data': violation.resource_data
+            }
+
+    def _retrieve(self):
+        """Retrieves the data for scanner.
+
+        Returns:
+            dict: User project relationship.
+        """
+        # This dictionary is the result of the scan.  The key
+        # is the user ID.  The value is a list of lists of ancestries.
+        project_ancestries_by_user = dict()
+        user_count = 0
+
+        start_time = time.time()
+
+        for user_email in user_emails(self.service_config):
+            user_count += 1
+
+            user_client = UserCloudResourceManagerClient(user_email,
```

Comment:
What is the reason for creating a separate `UserCloudResourceManagerClient` class?
I feel that after you enable a custom `credentials` to be passed into the classes in `cloud_resource_manager.py`, then you can just instantiate `CloudResourceManagerClient` directly, and it will have the correct user scope.  So, it's not necessary to have the `UserCloudResourceManagerClient` class?
```
for every user.....

user_scoped_credential = api_helpers.get_delegated_credential(user_email, SCOPES)
client = CloudResourceManagerClient(
    global_configs=inventory_configs,
    credentials=user_scoped_credential)
user_projects = client.get_projects()
user_project_ids = blah blah blah
```

Reply:
I did this because I was attempting higher cohesion.  Without the subclass, I would desire to put the ```project_ids``` and ```get_project_ancestry_as_resources``` methods in the ```CloudResourceManagerClient``` itself.  I'll try that and we'll see how it works.